### PR TITLE
IDENTITY-6509: Properly validate recovery endpoint webapp url before …

### PR DESCRIPTION
**Pull request by nuwandiw**
### Proposed changes in this pull request

- Recovery and self signup links will be shown as configured in identity.xml

### When should this PR be merged

immediately

### Follow up actions

- Create doc jira to update documents on ne configurations

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java [hunk 1]
```diff
@@ -942,4 +946,23 @@ public static boolean isSupportedByUserStore(UserStoreManager userStoreManager,
...
+     * This method creates a HttpClient instance with Protocol set to use SSLProtocolSocketFactory for https
+     *
+     * @param scheme
+     * @param host
+     * @return
+     */
+    public static HttpClient getHttpClientWithSSLSocketFactory(String scheme, String host) {
+        HttpClient client = new HttpClient();
+
+        if ("https".equals(scheme)) {
```
Comment (sagara-gunathunga) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java]: Shouldn't  we use a constant  here ?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/basicauth.jsp [hunk 2]
```diff
@@ -160,7 +162,15 @@
...
-            String url = identityMgtEndpointContext + "/recoverpassword.do?callback=" + Encode.forHtmlAttribute(urlEncodedURL);
+            HttpClient client = IdentityUtil.getHttpClientWithSSLSocketFactory(scheme, serverName);
+            String validationPointUrl = identityMgtEndpointContext + "/error.jsp";
+            HeadMethod head = new HeadMethod(validationPointUrl);
+            head.setFollowRedirects(false);
+
+            int status = client.executeMethod(head);
+            head.releaseConnection();
+            if (status == HttpURLConnection.HTTP_OK) {
+                String url = identityMgtEndpointContext + "/recoverpassword.do?callback=" + Encode.forHtmlAttribute(urlEncodedURL);
```
Comment (sagara-gunathunga) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/basicauth.jsp]: Better to wrap  url creation with method which implies  its purpose

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/basicauth.jsp [hunk 3]
```diff
@@ -160,7 +162,15 @@
                 identityMgtEndpointContext = IdentityUtil.getServerURL("/accountrecoveryendpoint", true, true);
             }
 
-            String url = identityMgtEndpointContext + "/recoverpassword.do?callback=" + Encode.forHtmlAttribute(urlEncodedURL);
+            HttpClient client = IdentityUtil.getHttpClientWithSSLSocketFactory(scheme, serverName);
```
Comment (IsuraD) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/basicauth.jsp]: Can we check the servlet context registered or not here, instead of executing a head call?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/basicauth.jsp (hunk 2 above)
Comment (nuwandiw) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/basicauth.jsp]: Done

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java [hunk 4]
```diff
@@ -61,13 +61,6 @@
 import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
 import org.xml.sax.SAXException;
 
-import javax.crypto.Mac;
```
Comment (IsuraD) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java]: Shall we revert these changes?
